test/integration/job/job_test.go

```diff
@@ -494,6 +494,439 @@ func TestJobPodFailurePolicy(t *testing.T) {
+					Ready:       ptr.To[int32](0),
+					Terminating: ptr.To[int32](0),
+				})
+				validateIndexedJobPods(ctx, t, clientSet, jobObj, podTermination.wantActiveIndexes, podTermination.wantCompletedIndexes, podTermination.wantFailedIndexes)
+			}
+			for i := range tc.wantConditionTypes {
+				validateJobCondition(ctx, t, clientSet, jobObj, tc.wantConditionTypes[i])
+			}
+			for i := range tc.wantJobFinishedNumMetric {
+				validateCounterMetric(ctx, t, metrics.JobFinishedNum, tc.wantJobFinishedNumMetric[i])
+			}
+			validateFinishedPodsNoFinalizer(ctx, t, clientSet, jobObj)
+		})
+	}
+}
+
+// TestSuccessPolicy_GetCompleteConditionAfterAllPodsAreTerminated tests handling of pod successful.
+// When Job matches rules, Job doesn't get both SuccessCriteriaMet and Complete.
+func TestSuccessPolicy_GetCompleteConditionAfterAllPodsAreTerminated(t *testing.T) {
+	defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, features.JobSuccessPolicy, true)()
+	closeFn, resetConfig, clientSet, ns := setup(t, "success-policy-get-complete")
+	defer closeFn()
+	ctx, cancel := startJobControllerAndWaitForCaches(t, resetConfig)
+	defer cancel()
+	resetMetrics()
+
+	jobObj, err := createJobWithDefaults(ctx, clientSet, ns.Name, &batchv1.Job{
+		Spec: batchv1.JobSpec{
+			Parallelism:    ptr.To[int32](5),
+			Completions:    ptr.To[int32](5),
+			CompletionMode: completionModePtr(batchv1.IndexedCompletion),
+			SuccessPolicy: &batchv1.SuccessPolicy{
+				Rules: []batchv1.SuccessPolicyRule{{
+					SucceededCount: ptr.To[int32](1),
+				}},
+			},
+		},
+	})
+	if err != nil {
+		t.Fatalf("Failed to create Job: %v", err)
+	}
+	validateJobPodsStatus(ctx, t, clientSet, jobObj, podsByStatus{
+		Active:      5,
+		Ready:       ptr.To[int32](0),
+		Terminating: ptr.To[int32](0),
+	})
+	validateIndexedJobPods(ctx, t, clientSet, jobObj, sets.New(0, 1, 2, 3, 4), "", nil)
+
+	// First pod from index 0 succeeded
+	if err = setJobPhaseForIndex(ctx, clientSet, jobObj, v1.PodSucceeded, 0); err != nil {
+		t.Fatalf("Failed tring to succeess pod with index 0")
+	}
+	validateJobPodsStatus(ctx, t, clientSet, jobObj, podsByStatus{
+		Active:      0,
+		Succeeded:   1,
+		Ready:       ptr.To[int32](0),
+		Terminating: ptr.To[int32](0),
+	})
+	validateIndexedJobPods(ctx, t, clientSet, jobObj, sets.New[int](), "0", nil)
+
+	if err = wait.PollUntilContextTimeout(ctx, 10*time.Millisecond, time.Minute, true, func(ctx context.Context) (bool, error) {
+		job, err := clientSet.BatchV1().Jobs(jobObj.Namespace).Get(ctx, jobObj.Name, metav1.GetOptions{})
+		if err != nil {
+			return true, err
+		}
+		if getJobConditionStatus(ctx, job, batchv1.JobSuccessCriteriaMet) != v1.ConditionTrue {
+			return true, nil
+		}
+		if getJobConditionStatus(ctx, job, batchv1.JobComplete) != "" {
+			t.Fatalf("Should not added Complete condition, yet")
+		}
+		return false, nil
+	}); err != nil {
+		t.Fatalf("Job should have only SuccessCriteriaMet, and should not have Complete")
```

Comment:
As discussed, there is no guarantee that the poll would catch the Job at this stage. It is possible that the following reconcile runs too fast and the Complete condition gets set.

Maybe we can just remove this entire test, as it's not adding much value compared to TestSuccessPolicy

Reply:
As I mentioned in https://github.com/kubernetes/kubernetes/pull/123412#discussion_r1516759021, I agree with removing `TestSuccessPolicy_GetCompleteConditionAfterAllPodsAreTerminated `.

Reply:
I remove it.